File: providers/aws/src/test/kotlin/cloudspec/aws/ec2/EC2SecurityGroupLoaderTest.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2;

import datagen.services.ec2.model.SecurityGroupGenerator;
import org.junit.Test;
import software.amazon.awssdk.services.ec2.model.DescribeSecurityGroupsRequest;
import software.amazon.awssdk.services.ec2.model.DescribeSecurityGroupsResponse;
import software.amazon.awssdk.services.ec2.model.SecurityGroup;

import java.util.function.Consumer;

import static org.junit.Assert.*;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doAnswer;

public class EC2SecurityGroupLoaderTest extends EC2LoaderTest {
    private final EC2SecurityGroupLoader loader = new EC2SecurityGroupLoader(getClientsProvider());

    @Test
    public void shouldLoadSecurityGroups() {
        var securityGroups = SecurityGroupGenerator.securityGroups();
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeSecurityGroupsRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeSecurityGroupsRequest.builder();
            builderLambda.accept(builder);
            assertTrue(builder.build().filters().isEmpty());
            return DescribeSecurityGroupsResponse.builder()
                                                 .securityGroups(securityGroups.toArray(new SecurityGroup[0]))
                                                 .build();
        }).when(getEc2Client()).describeSecurityGroups(any(Consumer.class));

        var resources = loader.getAll();

        assertNotNull(resources);
        assertEquals(securityGroups.size(), resources.size());
    }

    @Test
    public void shouldNotLoadSecurityGroupByRandomId() {
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeSecurityGroupsRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeSecurityGroupsRequest.builder();
            builderLambda.accept(builder);
            assertFalse(builder.build().filters().isEmpty());
            return DescribeSecurityGroupsResponse.builder()
                                                 .build();
        }).when(getEc2Client()).describeSecurityGroups(any(Consumer.class));

        var resource = loader.byId(SecurityGroupGenerator.securityGroupId());

        assertNotNull(resource);
    }

    @Test
    public void shouldLoadSecurityGroupById() {
        var securityGroups = SecurityGroupGenerator.securityGroups(1);
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeSecurityGroupsRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeSecurityGroupsRequest.builder();
            builderLambda.accept(builder);
            assertFalse(builder.build().filters().isEmpty());
            return DescribeSecurityGroupsResponse.builder()
                                                 .securityGroups(securityGroups.toArray(new SecurityGroup[0]))
                                                 .build();
        }).when(getEc2Client()).describeSecurityGroups(any(Consumer.class));

        var securityGroup = securityGroups.get(0);
        var resource = loader.byId(securityGroup.groupId());
        assertNotNull(resource);
    }
}

File: providers/aws/src/test/kotlin/cloudspec/aws/ec2/EC2SecurityGroupLoaderTest.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import arrow.core.None
import arrow.core.Some
import datagen.services.ec2.model.SecurityGroupGenerator.securityGroupId
import datagen.services.ec2.model.SecurityGroupGenerator.securityGroups
import io.mockk.every
import org.junit.Test
import software.amazon.awssdk.services.ec2.model.DescribeSecurityGroupsRequest
import software.amazon.awssdk.services.ec2.model.DescribeSecurityGroupsResponse
import java.util.function.Consumer
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class EC2SecurityGroupLoaderTest : EC2LoaderTest() {
    private val loader = EC2SecurityGroupLoader(clientsProvider)

    @Test
    fun `should load all resources`() {
        val securityGroups = securityGroups()

        every {
            ec2Client.describeSecurityGroups(any<Consumer<DescribeSecurityGroupsRequest.Builder>>())
        } answers {
            val builder = DescribeSecurityGroupsRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeSecurityGroupsRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertTrue(request.filters().isEmpty())

            DescribeSecurityGroupsResponse.builder()
                    .securityGroups(securityGroups)
                    .build()
        }

        val resources = loader.all(emptySet())
        assertNotNull(resources)
        assertEquals(securityGroups.size, resources.size)
    }

    @Test
    fun `should not load resource by random id`() {
        every {
            ec2Client.describeSecurityGroups(any<Consumer<DescribeSecurityGroupsRequest.Builder>>())
        } answers {
            val builder = DescribeSecurityGroupsRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeSecurityGroupsRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeSecurityGroupsResponse.builder()
                    .build()
        }

        val resourceOpt = loader.byId(emptySet(), securityGroupId())
        assertTrue(resourceOpt is None)
    }

    @Test
    fun `should load resource by id`() {
        val securityGroups = securityGroups()

        every {
            ec2Client.describeSecurityGroups(any<Consumer<DescribeSecurityGroupsRequest.Builder>>())
        } answers {
            val builder = DescribeSecurityGroupsRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeSecurityGroupsRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeSecurityGroupsResponse.builder()
                    .securityGroups(securityGroups)
                    .build()
        }

        val securityGroup = securityGroups[0]
        val resourceOpt = loader.byId(emptySet(), securityGroup.groupId())
        assertTrue(resourceOpt is Some<*>)
    }
}
